DESIGN.md

```diff
@@ -0,0 +1,364 @@
+```
+
+For more examples, check out the standard library for the example language in
+[`Language.Pirouette.Example.StdLib`]. The standard library also comes with a
+quasi-quoter of its own allowing you to leverage it in your code should you want
+it:
+
+```haskell
+myProgram :: PrtUnorderedDefs Ex
+myProgram = [progWithStdLib|
+  sum : List Integer -> Integer
+  sum =
+    foldr
+      @Integer
+      @Integer
+      (\(x : Integer) (y : Integer) . x + y)
+      0
+|]
+```
+
+### Plutus Intermediate Representation
+
+[plutus intermediate representation]: #plutus-intermediate-representation
+
+To run Pirouette on Plutus Intermediate Representation scripts, one has to use
+the [pirouette-plutusir library].
+
+[pirouette-plutusir library]: https://github.com/tweag/pirouette-plutusir
+
+Transformations
+---------------
+
+[transformations]: #transformations
+
+_This part is empty and should be documented._ Transformations live in
+[`Pirouette.Transformations`].
+
+Symbolic Engine
+---------------
+
+[symbolic engine]: #symbolic-engine
+
+The symbolic engine is split into three main parts: an SMT prover, as well as
+all the administrative work to communicate with it, Pirouette's _[symbolic
+evaluator]_, and Pirouette's _[symbolic prover]_. We describe these three
+aspects in the following sub-sections.
+
+### SMT prover
+
+As is classic in symbolic engines, Pirouette relies on SMT provers to do the
+hard work. These provers are used in two different ways. They help pruning
+branches of execution that are actually unreachable, and they help deciding
+whether the user's statement/questions actually hold.
+
+Pirouette relies on [smtlib-backend] for the communication with its SMT prover
+of choice, [Z3]. There are still a bunch of modules in Pirouette making the
+interface between the rest of the engine and [smtlib-backends], mainly to
+convert assumptions/constraints to the [SMT-LIB] language. These modules live in
+[`Pirouette.SMT`].
+
+[z3]: https://github.com/Z3Prover/z3
+[smtlib-backends]: https://github.com/tweag/smtlib-backends
+[smt-lib]: https://smtlib.cs.uiowa.edu/
+
+### Symbolic Evaluator
+
+[symbolic evaluator]: #symbolic-evaluator
+
+#### The symbolic evaluation monad — `SymEval`
+
+The heart of Pirouette's symbolic engine lives in [`Pirouette.Symbolic.Eval`].
+This module defines the `SymEval` monad providing all the tooling to make
+symbolic evaluation work. Here is the monad in question:
+
+```haskell
+newtype SymEval lang a = SymEval
+  { symEval ::
+      ReaderT
+        (SymEvalEnv lang)
+        (StateT (SymEvalSt lang) WeightedList)
+        a
+  }
+```
+
+Morally speaking, though, and forgetting the `lang` type parameters, this monad
+simply represent functions of type:
+
+```haskell
+SymEvalEnv -> SymEvalSt -> [(a, SymEvalSt)]
+```
+
+taking an environment and a state and returning a list of state, as well as
+potential return values. It returns a list because symbolic evaluation may
+branch and explore different paths.
```

Comment:
I ignore the `WeightedList`s indeed, but mostly because I can't really explain why they're there. I have a few vague ideas, but nothing clear enough to make it into this document. My impression is that a list would be fine and that `WeightedList`s are a case of early and over- engineering, but that makes me think that I missed something. For now, I added da71ec1 that says explicitly that we assume that they are lists. I'll open an issue to track this missing point.